### ISSUE TITLE
bug: createPersistentStore should infer store type from decode return type

### ISSUE THREAD

**Issue by droganov** [nanostores/persistent#5]
In this example createPersistentStore blocks legitimate value type because it's hardcoded to `<Value extends string | undefined>`

I think it should infer value type from decode result with fallback to `string | undefined` in order to actually support  the options

```js
import { createPersistentStore } from '@nanostores/persistent'
import { useStore } from 'nanostores/react'

type Settings = {
  skipRules?: boolean
}

const settingsStore = createPersistentStore<Settings>(
  'settings',
  {},
  {
    encode(value: Settings): string {
      return JSON.stringify(value)
    },
    decode(value: string): Settings {
      try {
        return JSON.parse(value) as Settings
      } catch (error) {
        // todo: log error
        return {}
      }
    },
  }
)

export const useSettings = (): Settings => useStore(settingsStore)

```

<img width="789" alt="image" src="https://user-images.githubusercontent.com/1375299/132105090-f231fc0e-6bd5-403f-bd7a-30c853f7cd71.png">

**Comment by droganov** [comment]
The fix would b something like this:
```js
export function createPersistentStore<Value = string | undefined>(
  name: string,
  initial?: Initial<Value> | undefined,
  opts?: PersistentOptions<Value>
): WritableStore<Value>

type Initial<Value> = PersistentOptions<Value>[decode] extends undefined ? string : Value

export interface PersistentOptions<Value> {
  encode(value: Value): void
  decode(value: string | null): Value
  listen?: boolean
}
```

**Comment by droganov** [comment]
This way we can handle both cases:
<img width="982" alt="image" src="https://user-images.githubusercontent.com/1375299/132122905-d6b1df4d-7c61-409e-9066-8e8fe18161aa.png">

<img width="721" alt="image" src="https://user-images.githubusercontent.com/1375299/132122929-89c35421-d92e-484e-8249-e5ce5cfa2c42.png">

**Comment by ai** [comment]
Yeap. The plan looks good.

I will try to implement it tomorrow.

**Comment by ai** [comment]
Fixed e7d42dd

**Comment by ai** [comment]
Released in 0.3.1
